ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,284 @@
+package ring
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	totalTokensCount         = 1 << 32
+	optimalTokensPerInstance = 1 << 9
+	maxZonesCount            = 1 << 3
+)
+
+var (
+	instanceIDRegex          = regexp.MustCompile(`^(.*)-(zone-[a-z])-(\d+)$`)
```

Comment:
Why do we require "-zone-X-" in the ID at all? We just need to parse the ID cause the zone will be an information in input. If not strictly required, I would relax this, so that there's no assumption on the format of the zone in the instance ID.